File: CALISTA-MATLAB/subfunctions/CALISTA_clustering.m
function [final_results] = CALISTA_clustering(DATA,INPUTS,Results,varargin )


% CALISTA is a novel computational method for Clustering And 
% Lineage Inference through Single Cell Transcriptomics Analysis

if nargin <3
    error('Not enough input variables.')
end

mRNA_all=DATA.totDATA;
k_new=DATA.Parameters.sets;
expected_clusters=Results.expected_clusters;

if length(expected_clusters)>1
    error('Please input only one value for the expected number of clusters')
end

if INPUTS.use_drop_prob_in_clustering
    p_mat_ma=DATA.Parameters.P;
    opt_lambda=DATA.opt_lambda;
    % invert the probability matrix  %NOT LOG AS IN TRANSITION GENES!
    X=p_mat_ma';
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % Define dropout probability
    epsilon=min(min(X(X>0)));
    mRNA_counts=[0 1:200];
    P_temp=exp(-opt_lambda*mRNA_counts)';
    P=diag([1; 1-P_temp(2:end)]);
    P(:,1)=P_temp;
    Z_temp=X*P;
    Z_temp(Z_temp==0)=epsilon;  %%%%%%% SET ZEROS TO A SMALL VALUE
    Z_temp=log(Z_temp)';
else
    Z_temp=DATA.Parameters.logP;
end


%% Default values
% run optimiziation
optimize=true;
% initial population
in_pop=false;
% number of runs
loops=50;
% run in parallel
run_parallel=false;
% cluster algorithm ('greedy_cabsel',greedy_maxdiff', 'cabsel_sabec')
algorithm='greedy_cabsel';
% number of maximum iterations
max_iter=100;
% number of cells and number of genes
[nvars, n_genes]=size(mRNA_all);
nVarargs=length(varargin);
% plot
my_plot=true;
% clustering
my_cluster='kmedoids';

for k=1:2:nVarargs
    
    
    % check if user only wants to get the optimal rate constants for a
    % given assignment
    if strcmp(varargin{k},'get_k')
        % 1 assignement to calculate optimal rate constants
        as_get_k=varargin{k+1};
        % do not optimize
        optimize=false;
        % therefore number of runs =1
        loops=1;
        % if more than one population throw error
        if size(as_get_k,1)>1
            error('Please provide only 1 initial population in order to calculate the rate constants wihout optimization')
        end
        
    end
    
    % check if user has userdefined initial populations
    if strcmp(varargin{k},'set_initial_pop')
        in_pop=true;
        as_in=varargin{k+1};
        
    end
    
    
    % number of runs (default =50)
    if strcmp(varargin{k},'pop_size')
        loops=varargin{k+1};
    end
    
    % number of iterations for each run (default =100)
    if strcmp(varargin{k},'max_iterations')
        max_iter=varargin{k+1};
    end
    
    
    % run function in parallel (default false)
    if strcmp(varargin{k},'parallel')
        run_parallel=varargin{k+1};
    end
    
    % choose algorithm
    if strcmp(varargin{k},'algorithm')
        algorithm=varargin{k+1};
    end
    
    % choose algorithm
    if strcmp(varargin{k},'cluster')
        my_cluster=varargin{k+1};
    end
    
    if strcmp(varargin{k},'use_imputed_data')
        mRNA_all=varargin{k+1};
        [nvars, n_genes]=size(mRNA_all);
        
    end
end

% if no optimization just run for 1 population
if optimize==false
    as_all=as_get_k;
    % if expected_clusters is empty get clusters from the population
    if isempty(expected_clusters)
        expected_clusters=length(unique(as_all));
    end
elseif length(expected_clusters)==1
    % if optimization is true sample random numbers between 1 and expected clusters
    if INPUTS.data_type==5
        seed=round(nvars/2000*expected_clusters);
    else
        seed=round(nvars/2*expected_clusters);
    end
    rng(seed) % For reproducibility
    as_all=randi([1 expected_clusters],loops,nvars);
    
end

% if user provides initial poplation and at the same time wants to
% calculate the optimal constants, throw warning
if in_pop==true
    
    if optimize==false
        warning('No optimization is carried out, since option "get_k" was set')
        
    else
        
        % if length of initial population is smaller than number of runs defined, throw warning
        if size(as_in,1)<loops
            
            loops=size(as_in,1);
            warning('Number of iteratios was set according to size of the initial population')
        end
        
        as_all=as_in;
        
    end
    
end

% define variables used in the function greedy_cabsel
in_population=zeros(loops,nvars);
sum_prob_tot=zeros(1,loops);
opt_idx=struct('run',[]);
opt_idx_a(1:loops)=opt_idx;

% check option parallel was set
if run_parallel==false
    p=0;
else
    
    aaa=gcp('nocreate');
    % if no parpool was started
    if isempty(aaa)
        parpool;
        aaa=gcp('nocreate');
    end
    % number of workes in parfor
    p=aaa.NumWorkers;
    p=p-1;  % number of CPUs used for parallel execution (minus 1 for system stability)
end
%% RUN Greedy Algorithm
display=1;
[my_results ] = greedy_cabsel( as_all,Z_temp,k_new,mRNA_all,n_genes,max_iter,nvars,...
    optimize,opt_idx_a,p,sum_prob_tot,in_population,loops,expected_clusters,algorithm,display);

my_results.cluster=expected_clusters;
all_my_results.all=my_results;

if optimize==true
    
    switch INPUTS.step
        case 1
            population=my_results.population;
            % consensus
            fprintf('\nConstructing consensus matrix...\n');
            consensus=get_consensus(population,p);
        case 2
            
            
            if INPUTS.data_type==5
                fprintf('\nBest cell assignment...\n');
                idx=my_results.best';
            else
                fprintf('\nPerforming clustering based on consensus matrix...\n');
                population=my_results.population;
                % consensus
                fprintf('\nConstructing consensus matrix...\n');
                consensus=get_consensus(population,p);
                data_2_cluster=consensus;
                
                switch my_cluster
                    case 'kmedoids'
                        fprintf('\nKmedois...\n');
                        stream = RandStream('mrg32k3a');  % Random number stream
                        options = statset('UseParallel',run_parallel,'UseSubstreams',run_parallel,...
                            'Streams',stream);
                        idx=kmedoids(data_2_cluster,expected_clusters,'Algorithm','pam','Replicates',5,'Options',options);
                    case 'hierarchical'
                        fprintf('\nHierarchical clustering...\n');
                        Z = linkage(data_2_cluster,'complete');%'average','correlation'); %'complete);
                        idx=cluster(Z,'maxclust',expected_clusters);
                end
            end
   
    
    display=0;
    [my_results_1 ] = greedy_cabsel(idx',Z_temp,k_new,mRNA_all,n_genes,max_iter,nvars,...
        false,opt_idx_a,p,sum_prob_tot,in_population,1,expected_clusters,algorithm,display);
    all_my_results.all.clusterprobabilities=my_results_1.clusterprobabilities;
    all_my_results.all.idx=idx;
    end
    if exist('consensus','var')
    all_my_results.all.consensus=consensus;
    end
end

final_results.all=all_my_results;
end

File: CALISTA-R/calista/R/CALISTA_clustering.R

#' A calista Function
#'
#' @param mRNA_all,log_p_mat_ma,K_new,expected_clusters,...
#' @keywords calista
#' @export
#' @examples
#' CALISTA_clustering()






CALISTA_clustering <- function(mRNA_all,log_p_mat_ma,K_new,expected_clusters,...){

  #my_results_c=CALISTA_clustering(DATA$totDATA,Parameters$Parameters[[3]],
  #Parameters$Parameters[[1]],Results$expected_clusters,
  #'parallel',parallel,'cluster',Cluster,'algorithm',algorithm)
  #test
   # mRNA_all=DATA$totDATA
   # log_p_mat_ma=Parameters$Parameters[[3]]
   # K_new=Parameters$Parameters[[1]]
   # expected_clusters=Results$expected_clusters
   # varagin=list('parallel',parallel,'cluster',Cluster,'algorithm',algorithm)

  varagin=list(...)
  if(nargs()<4){
    stop('Not enough input variables.')
  }
  if (length(expected_clusters)>1){
    stop('Please input only one value for the expected number of clusters')
  }
  ##default values
  #fun optimizization
  optimize=TRUE
  #initial population
  in_pop=FALSE
  #number of runs
  loops=50 # change to 50 later !!!!!!!!!
  #run in parallel
  run_parallel=FALSE
  #cluter algorithm
  algorithm='greedy_cabsel'
  #num of maximum iterations
  max_iter=100
  #number of cells and number of genes
  nvars=nrow(mRNA_all)
  n_genes=ncol(mRNA_all)
  nVarargs=length(varagin)
  #plot
  my_plot=TRUE
  #clustering
  my_cluster='kmedoids'
  as_all=list()
  as_get_k=NULL
  for (k in seq(from=1,to=nVarargs,by=2)) {
    if(varagin[[k]]=='get_k'){
      as_get_k=varagin[[k+1]]
      optimize=FALSE
      loops=1
      if(nrow(as_get_k)>1){
        stop('Please provide only 1 initial population in order to calculate the rate constants wihout optimization')
      }
    }
    if(varagin[[k]]=='set_initial_pop'){
      in_pop=TRUE
      as_in=varagin[[k+1]]
    }
    if(varagin[[k]]=='pop_size'){
      loops=varagin[[k+1]]
    }
    if(varagin[[k]]=='max_iteration'){
      max_iter=varagin[[k+1]]
    }
    if(varagin[[k]]=='parallel'){
      run_parallel=varagin[[k+1]]
    }
    if(varagin[[k]]=='algorithm'){
      algorithm=varagin[[k+1]]
    }
    if(varagin[[k]]=='cluster'){
      my_cluster=varagin[[k+1]]
    }
  }
  if(optimize==FALSE){
    as_all=as_get_k
    if(is.null(expected_clusters)){
      expected_clusters=length(unique(as_all))
    }
  }else if(length(expected_clusters==1)){
    set.seed(round(nvars/20*expected_clusters))
    as_all=matrix(sample(1:expected_clusters,loops*nvars,replace = TRUE),nrow = loops)
  }

  ### if user provides initial poplulation and at the same time want to
  ### calculate the optimal constants, throw warning
  if (in_pop==TRUE){
    if(optimize==FALSE){
      stop('No optimization is carried out, since option "get_k" was set')
    }else{
      if(nrow(as_in)<loops){
        loops=nrow(as_in)
        warning('Number of iteratios was set according to size of the initial population')
        as_all=as_in
      }
    }
  }
  #define variable used in the function greedy_cabsel
  in_population=matrix(0,loops,nvars)
  sum_prob_tot=matrix(0,1,loops)
  opt_idx=list(run=NULL)
  opt_idx_a=list()
  opt_idx_a[1:loops]=opt_idx
  p=NULL
  if(run_parallel==FALSE){
    p=1
  }else{
    p=detectCores()-1
    # if(is.null(p)){
    #   cl=makeCluster(6)
    #   registerDoParallel(cl)
    # }
    # p=getDoParWorkers()
  }
  # run greedy algorithm
  display=1
  # in_population=matrix(0,loops,nvars)
  # sum_prob_tot=matrix(0,1,loops)
  # opt_idx=list(run=NULL)
  # opt_idx_a=list()
  # opt_idx_a[1:loops]=opt_idx
  # p=NULL
  # if(run_parallel==FALSE){
  #   p=1
  # }else{
  #   p=detectCores()-1
  #   # if(is.null(p)){
  #   #   cl=makeCluster(6)
  #   #   registerDoParallel(cl)
  #   # }
  #   # p=getDoParWorkers()
  # }
  #num_clusters=length(expected_clusters)

  #write greedy_cabsel to get my_results here
  #here notice as_all
  #source('greedy_cabsel.R')

  my_results=greedy_cabsel_ff(as_all,log_p_mat_ma,K_new,mRNA_all,n_genes,max_iter,nvars,optimize,opt_idx_a,p,sum_prob_tot,in_population,loops,expected_clusters,algorithm,display)
  my_results$cluster=expected_clusters
  population=my_results$population
  #consensus
  #source('get_consensus.R')
  consensus=get_consensus(population) # implement get_consensus here
  all_my_results=list()
  all_my_results$all=my_results
  i=0
  final_results=list()
  #source('greedy_cabsel.R')

  if(optimize==TRUE){
    switch (my_cluster,
            # 'kmedoids' = {idx=pam(consensus,expected_clusters)$clustering},
            'kmedoids' = {dissimilarity_matrix=1-consensus/loops
            idx=as.integer(Cluster_Medoids(dissimilarity_matrix,clusters = expected_clusters,
                                           distance_metric = "euclidean",seed = 1)$clusters)},
            #'kmedoids' =
            # print(dim(consensus))
            #  idx=Cluster_Medoids(consensus,clusters = expected_clusters, distance_metric = "euclidean",seed = 10)$clusters
            #  idx=as.integer(idx)
            #  print(idx)
            #  str(idx)},
            'hierarchical'={Z=hclust(dist(scale(consensus)),method = 'complete')
            idx=cutree(Z,k=expected_clusters)}
    )
    }else{idx=as_all}
    display=0


    my_results_1=greedy_cabsel_f(t(idx),log_p_mat_ma,K_new,mRNA_all,n_genes,max_iter,nvars,FALSE,p,
                                 sum_prob_tot,population,1,expected_clusters,algorithm,display,opt_idx_a) #here,loops can only be one

    all_my_results$all$clusterprobabilities=my_results_1$clusterprobabilities
    all_my_results$all$idx=idx
    all_my_results$all$cell_prob = my_results_1$cell_prob
    # my_distance=my_results_1$cell_prob#matrix(0,nvars,expected_clusters)
    # for (m in 1:expected_clusters) {
    #   aaa=which(idx==m)
    #   my_distance[aaa,]=abs(rep( my_results_1$cell_prob[aaa,m],1*expected_clusters)- my_results_1$cell_prob[aaa,])   #check heres
    # }
    all_my_results$all$distance =  my_results_1$distance
    all_my_results$all$param_idx = my_results_1$param_idx
    all_my_results$all$consensus = consensus



  final_results$all=all_my_results
  return(final_results)
  # return(idx)
}


















